controllers/machine.go

```diff
@@ -82,45 +93,92 @@ func (mrc *machineReconcileContext) addFinalizer() error {
 	return nil
 }
 
-func (mrc *machineReconcileContext) ensureDependencies() error {
-	hardware, err := mrc.ensureHardware()
-	if err != nil {
-		return fmt.Errorf("ensuring hardware: %w", err)
-	}
+func (mrc *machineReconcileContext) isHardwareReady(hw *tinkv1.Hardware) bool {
+	return hw.Spec.Metadata.State == inUse && hw.Spec.Metadata.Instance.State == provisioned
+}
+
+type errRequeueRequested struct{}
+
+func (e *errRequeueRequested) Error() string {
+	return "requeue requested"
+}
+
+func (mrc *machineReconcileContext) ensureTemplateAndWorkflow(hw *tinkv1.Hardware) (*tinkv1.Workflow, error) {
+	wf, err := mrc.workflow()
+
+	switch {
+	case apierrors.IsNotFound(err):
+		if err := mrc.ensureTemplate(hw); err != nil {
+			return nil, fmt.Errorf("failed to ensure template: %w", err)
+		}
+
+		if err := mrc.createWorkflow(hw); err != nil {
+			return nil, fmt.Errorf("failed to create workflow: %w", err)
+		}
 
-	if err := mrc.ensureTemplate(hardware); err != nil {
-		return fmt.Errorf("ensuring template: %w", err)
+		return nil, &errRequeueRequested{}
+	case err != nil:
+		return nil, fmt.Errorf("failed to get workflow: %w", err)
+	default:
 	}
 
-	if err := mrc.ensureWorkflow(hardware); err != nil {
-		return fmt.Errorf("ensuring workflow: %w", err)
+	s := wf.GetCurrentActionState()
+	if s == tinkv1.WorkflowStateFailed || s == tinkv1.WorkflowStateTimeout {
+		mrc.log.Info("current action is in a failed or timed out state: %v", wf.GetCurrentAction())
 	}
 
-	return nil
+	return wf, nil
 }
 
-func (mrc *machineReconcileContext) markAsReady() error {
-	mrc.tinkerbellMachine.Status.Ready = true
+//
+func (mrc *machineReconcileContext) Reconcile() error {
+	defer func() {
+		// make sure we do not create orphaned objects.
+		if err := mrc.addFinalizer(); err != nil {
+			mrc.log.Error(err, "error adding finalizer")
+		}
+	}()
 
-	if err := mrc.patch(); err != nil {
-		return fmt.Errorf("patching machine with ready status: %w", err)
+	hw, err := mrc.ensureHardware()
+	if err != nil {
+		return fmt.Errorf("failed to ensure hardware: %w", err)
 	}
 
-	return nil
-}
+	if !mrc.isHardwareReady(hw) {
+		wf, err := mrc.ensureTemplateAndWorkflow(hw)
 
-func (mrc *machineReconcileContext) Reconcile() error {
-	// To make sure we do not create orphaned objects.
-	if err := mrc.addFinalizer(); err != nil {
-		return fmt.Errorf("adding finalizer: %w", err)
+		switch {
+		case errors.Is(err, &errRequeueRequested{}):
+			return nil
+		case err != nil:
+			return fmt.Errorf("ensure template and workflow returned: %w", err)
+		}
+
+		if !mrc.lastActionStarted(wf) {
+			return nil
+		}
+
+		hw.Spec.Metadata.State = inUse
+		hw.Spec.Metadata.Instance.State = provisioned
+
+		if err := mrc.patchHardware(hw); err != nil {
+			return fmt.Errorf("failed to patch hardware: %w", err)
+		}
 	}
 
-	if err := mrc.ensureDependencies(); err != nil {
-		return fmt.Errorf("ensuring machine dependencies: %w", err)
+	mrc.tinkerbellMachine.Status.Ready = true
+
+	return mrc.patch()
```

Comment:
The defered `mrc.addFinalizer()` also does a [mrc.patch()](https://github.com/jacobweinstock/cluster-api-provider-tinkerbell/blob/94aa1e08f70e21cde9537c20161bbecb8e871195/controllers/machine.go#L89).

To avoid multiple patch calls can we do this in one place?

Reply:
Sounds good!